rfcs/0000-nixpkgs-ifd.md

```diff
@@ -0,0 +1,182 @@
+---
+
+# Summary
+[summary]: #summary
+
+Many of us want "import from derivation" \[hereafter, IFD\] be allowed in Nixpkgs.
+IFD would be terrible for `hydra.nixos.org`, though,
+The solution is to cut the Gordian knot: Allow IFD in Nixpkgs while still (effectively) prohibiting it in CI.
+
+# Motivation
+[motivation]: #motivation
+
+Nixpkgs, along with every other distro, also faces a looming crisis: new open source software is increasingly not intended to be packaged by distros at all.
+Many languages now support very large library ecosystems, with dependencies expressed in a language-specific package manager.
+To this new generation of developers, the distro (or homebrew) is a crufty relic from an earlier age to bootstrap modernity, and then be forgotten about.
+
+Right now, to deal with these packages, we either convert by hand, or commit lots of generated code into Nixpkgs.
+But I don't think either of those options is healthy or sustainable.
+The problem with the first is sheer effort; we'll never be able to keep up.
+The problem with the second is bloating Nixpkgs but more importantly reproducability: If someone wants to update that generated code it is unclear how.
+All these mean that potential users coming from this new model of development find Nix / Nixpkgs cumbersome and unsuited to their needs.
+
+The solution *outside* of Nixpkgs is today is "import from derivation", i.e. building code in Nix to be consumed at eval time.
+Many institutional users of Nix use this.
+But while the practice is banned in Nixpkgs, those efforts are not very coordinated, and the `lang2nix` ecosystem has a hard time getting off the ground.
+
+I am *not* arguing that IFD is the best possible solution.
+But it's the one we've got to day, and long term alternatives, like RFC #92, face *significant* hurdles in being ergonomic and integrating with current idioms in Nixpkgs -- e.g. the `meta` on every derivation from `mkDerivation`.
+In the spirit of learning to walk before learning to run, and beginning to acknowledge addresses these problems, we are best-serviced by getting IFD in Nixpkgs as a first-gen solution as soon as possible.
+The only barrier then is addressing eval resource usage costs.
+
+# Detailed design
+[design]: #detailed-design
+
+## Nixpkgs
+
+1. Add a new `enableImportFromDerivation` config parameter to Nixpkgs.
+   When it is `false`, anything using IFD must be disabled so that a regular evaluation like we do today succeeds.
+
+2. Add a new `allImportedDerivations` top-level attribute.
+   This *must* be buildable with `enableImportFromDerivation = false`.
+   It *must* have in its run-time closure any derivation output that Nixpkgs with `enableImportFromDerivation = true` imports.
+   \(CI will verify these conditions as described in the next subsection.\)
+
+3. Any code vendored in Nixpkgs *must* correspond to code produced in an imported derivation, so the code can be mechanistically re-vendored.
```

Comment:
We shouldn't vendor any more than today. And in fact we should vendor less: instead of vendoring all of hackage, we should just vendor enough to build `cabal2nix`, and bootstrap everything else wit IFD from that.

But I don't want to put in the RFC what we should and shouldn't vendor, just establish the principles by which vendoring is allowed today.

Even if we kept all Hackage vendored as today, it would be still better if you could just update a `fetchgit` of the relevant from data from Nixpkgs and then regenerate everything purely.

Reply:
> And in fact we should vendor less

but only when we can remove parts of the vendored files which would require the final nix user to have IFD enabled which is not going to happen anytime soon if I understood it correct. And that is my main point: this will only come in really useful if we don't need any vendor files anymore.

Reply:
@SuperSandro2000 I'm a bit confused?

> which is not going to happen anytime soon if I understood it correct

Yes, I do say I don't want to immediately crate a big dust-up making `haskellPackages` IFD only.

What I rather do is go in the other direction. There are many would-be things like `haskellPackages` that don't get merged because we don't have an appetite for more generated code. Your response to @sternenseemann's wanted to add a second stable package set, https://github.com/NixOS/nixpkgs/pull/138407#issuecomment-924286406, is in fact a prefect example of this! :)! Let's go those things out of limbo by making them use IFD instead of vendoring, and *them* pioneer the new IFD approach. 

I hope that well become a success, and we will then be open to using IFD for things that are vendored today, like the default `haskellPackages`.

Reply:
One prospect could in fact be to remove every haskell package that is marked as broken presently from `hackage-packages.nix` since those are never touched by CI anyways. We have been hesitant to do this so far, because some of those tend to fix themselves after a while or are really easy to unbreak, so removing them ruins the contributing experience.

Requiring local users to enable ifd to (try) to build these could actually be a fairly decent option.